### PR TITLE
First implementation

### PR DESCRIPTION
**Pull request by marcopiraccini**
First implementation. 
This is configured with:
```
{
    url: 'amqp://localhost',
    generateExchange: true,
    exchanges: [{
       name: 'test-exchange',
       targetUrl: 'http://localhost:3042/target'
   }]
}
```

### CODE REVIEW

File: lib/schema.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    ...schema.schema.properties,
+    module: { type: 'string' },
+    rabbitmq: {
+      type: 'object',
+      properties: {
+        url: { type: 'string' },
+        exchange: { type: 'string' },
+        routingKey: { type: 'string' },
+        targetUrl: { type: 'string' }
+      },
```
Comment (mcollina) [inline lib/schema.js:21]: I think we should support an array of those.

File: lib/plugin.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+      await request(targetUrl, {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json'
+        },
+        body: JSON.stringify({ data: message })
+      })
+    } catch (err) {
+      app.log.error(err)
+      throw new Error('Failed to send message to targetUrl')
```
Comment (mcollina) [inline lib/plugin.js:26]: This would never happen. In case you need to `nack` the message and it should automatically be retried

**Review by mcollina — CHANGES_REQUESTED** [review]
This should also expose routes to publish messages on those exchanges

File: lib/schema.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  properties: {
+    ...schema.schema.properties,
+    module: { type: 'string' },
+    rabbitmq: {
+      type: 'object',
+      properties: {
+        url: { type: 'string' },
+        exchange: { type: 'string' },
+        routingKey: { type: 'string' },
+        targetUrl: { type: 'string' }
```
Comment (mcollina) [inline lib/schema.js:20]: What method is going to be used? additional headers would also be required.

**Review by mcollina — CHANGES_REQUESTED** [review]

File: lib/schema.js (hunk 3 above)
Comment (marcopiraccini) [reply lib/schema.js]: We use `POST` in this PR.

File: lib/plugin.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            routingKey: { type: 'string' },
+            message: { type: 'string' }
+          }
+        }
+      },
+      handler: async (req, reply) => {
+        const { exchange, routingKey, message } = req.body
+        app.log.info(`Publishing to exchange ${exchange} with routing key ${routingKey}`)
+        await rabbit.publish(exchange, message, routingKey)
+        reply.send({ status: 'ok' })
```
Comment (mcollina) [inline lib/plugin.js:59]:
```suggestion
        return { status: 'ok' })
```

File: lib/plugin.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+  app.post('/publish',
+    {
+      schema: {
+        body: {
+          type: 'object',
+          properties: {
+            exchange: { type: 'string' },
+            routingKey: { type: 'string' },
+            message: { type: 'string' }
```
Comment (mcollina) [inline lib/plugin.js:51]: I think we should put exchange and routingkey as a routing parameters

File: lib/plugin.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+      app.log.info(`Routing key: ${routingKey}`)
+    }
+    await rabbit.listen(name, routingKey, async (msg) => {
+      const message = msg?.content?.toString()
+      app.log.debug('Received', message)
+      try {
+        await request(targetUrl, {
+          method: 'POST',
+          headers: {
+            'Content-Type': 'application/json'
```
Comment (mcollina) [inline lib/plugin.js:31]: Content type should be json unless it's specified inside the headers map (or the headers map as a json content type as default.

File: lib/rabbitmq.js [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    const channel = await this.#getChannel()
+    await this.#checkExchange(exchange)
+    const { queue } = await channel.assertQueue('', { exclusive: true })
+    await channel.bindQueue(queue, exchange, routingKey)
+    await channel.consume(queue, async (msg) => {
+      try {
+        await consumer(msg)
+      } catch (err) {
+        logger.error('Failed to consume message', err)
+        channel.nack(msg)
```
Comment (mcollina) [inline lib/rabbitmq.js:93]:
```suggestion
        channel.nack(msg)
        return
```

File: lib/rabbitmq.js [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+      }
+      throw err
+    }
+  }
+
+  async listen (exchange, routingKey, consumer) {
+    const logger = this.#logger
+    const channel = await this.#getChannel()
+    await this.#checkExchange(exchange)
+    const { queue } = await channel.assertQueue('', { exclusive: true })
```
Comment (mcollina) [inline lib/rabbitmq.js:86]: the queue name and exclusive should be configurable. Also `durable` should be there.

File: lib/rabbitmq.js [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+      await channel.assertExchange(exchange, 'fanout', { durable: false })
+      return
+    }
+
+    const logger = this.#logger
+    if (!this.#checkExchangeChannel) {
+      this.#checkExchangeChannel = await this.#connection.createChannel()
+    }
+    this.#checkExchangeChannel.on('error', (err) => {
+      logger.error('Failed to check exchange', err)
```
Comment (mcollina) [inline lib/rabbitmq.js:68]:
```suggestion
      logger.error({ err }, 'Failed to check exchange')
```

File: lib/rabbitmq.js [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+      this.#connection = null
+      this.#channel = null
+      this.#checkExchangeChannel = null
+    }
+    try {
+      this.#connection = await amqp.connect(url)
+    } catch (err) {
+      throw new Error(`Connection failed to ${url}`)
+    }
+    logger.info('Connected to RabbitMQ', url)
```
Comment (mcollina) [inline lib/rabbitmq.js:35]:
```suggestion
    logger.info('Connected to RabbitMQ %s', url)
```

File: lib/rabbitmq.js [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+      }
+      channel.ack(msg)
+    }, { noAck: false })
+    logger.info('Registered message consumer')
+  }
+
+  async publish (exchange, message, routingKey = '') {
+    await this.#checkExchange(exchange)
+    const logger = this.#logger
+    logger.info(message, 'Publishing message')
```
Comment (mcollina) [inline lib/rabbitmq.js:103]: message might be big and leaking private info, likely it should not be logged.

File: lib/rabbitmq.js (hunk 7 above)
Comment (marcopiraccini) [reply lib/rabbitmq.js:93]: Fixed

**Review by mcollina — CHANGES_REQUESTED** [review]
Could you add a test with two instances listening to the same exchange and queue? They should deliver the message only once.

**Comment by marcopiraccini** [comment]
> Could you add a test with two instances listening to the same exchange and queue? They should deliver the message only once.

It's already done here: https://github.com/platformatic/rabbitmq-hooks/blob/first-implementation/test/rabbitmq.test.js#L160

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**